전남대 BE_강명덕_2주차 과제 (3단계)

Profile-exe opened the pull request:
## 코드를 작성하며 어려웠던 점
- `Argument Resolver`라는 유용한 것을 알게 되었습니다. 하지만 처음 접하는 부분이라 강의 자료에 힌트로 제공되는 부분이 있어도 구현에 제법 시간이 걸렸습니다.

## 코드 리뷰 시, 멘토님이 중점적으로 리뷰해줬으면 하는 부분
- 위시리스트 추가 로직에서 기존에 상품이 존재하는지를 분기를 두어 확인했습니다. 추가 요청 대상인 상품이 이미 존재한다면 기존의 수량에서 요청된 수량만큼 더 증가시키는 방식으로 구현했습니다. 
  
  이러한 로직을 리파지토리에서 하나의 메서드로 처리할 수 있을텐데 서비스에 구현하는게 올바른 방식인지 궁금합니다. 제 생각에는 서비스에서 비즈니스 로직을 책임지고 리파지토리는 데이터를 조회, 조작하는 책임만 있다고 생각해 서비스에서 분기를 두고 리파지토리에는 각 분기에 해당하는 메서드를 따로 만들었습니다.

Profile-exe committed:
refactor(MemberRepository): 중복 검증로직 제거

`deleteMember`를 호출하는 `MemberService`에서 `validateMemberExistsById`로 존재여부를 이미 검증하고있다. 그래서 `deleteMember`메서드에 중복 로직이 존재해 이를 제거

Profile-exe committed:
docs(README.md): Step3 구현할 기능 목록 작성

Profile-exe committed:
feat(WishList): 위시 리스트 엔티티 추가

Profile-exe committed:
feat(WishListErrorCode): 위시 리스트 에러코드 추가

Profile-exe committed:
feat: 위시 리스트 비즈니스 예외 추가

Profile-exe committed:
feat(WishListRepository): 위시 리스트 리파지토리 추가

Profile-exe committed:
feat(WishListService): 위시 리스트 서비스 추가

Profile-exe committed:
feat(WishListResDto): 위시 리스트 정보를 반환하는 DTO 추가

Profile-exe committed:
feat(WishListReqDto): 위시 리스트 추가/수정 요청하는 DTO 추가

Profile-exe committed:
feat(LoginMember): 토큰에서 회원을 추출하는 argumentResolver 추가

Profile-exe committed:
feat(JwtInterceptor): 인가를 위한 인터셉터 추가

`WebConfig`에서 제외한 엔드포인트가 아니라면 토큰이 없을 때 접근 불가

Profile-exe committed:
feat(JwtTokenProvider): 인터셉터에 사용할 토큰 유효성 검증 메서드 추가

Profile-exe committed:
config(WebConfig): 인터셉터 및 argumentResolver 등록

로그인 및 회원가입은 토큰 없이도 가능하도록 구현

Profile-exe committed:
feat(WishListController): 위시 리스트 API 컨트롤러 추가

Profile-exe committed:
chore: 위시 리스트 테이블 및 데이터 추가

Profile-exe committed:
refactor(JwtInterceptor): 불필요한 throws 구문 제거

Profile-exe committed:
docs(README.md): 요청, 응답 예시 코드블럭 언어 수정

hocaron commented on src/main/java/gift/auth/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,45 @@
+    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest,
+            WebDataBinderFactory binderFactory) {
+
+        String token = webRequest.getHeader("Authorization");
```
다른 부분에 적용하신 것처럼 이곳도 HttpHeaders.AUTHORIZATION 를 사용하면, 작업 중에 문자열이 틀릴 가능성을 줄일 수 있을 것 같습니다😀

Profile-exe replied:
아 이 부분을 신경썼어야 했는데.. 그러지 못했네요. 꼼꼼히 봐주셔서 정말 감사합니다. 바로 반영했습니다. 

[cf0192bb](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/407/commits/cf0192bb9677a400f75c2781cd53a8c4301ff6a9) 커밋에 반영했습니다.

hocaron commented on src/main/java/gift/config/JwtInterceptor.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+
+        String token = request.getHeader(HttpHeaders.AUTHORIZATION);
+
+        if (token == null || token.length() <= BEARER_PREFIX_LENGTH) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
+        }
+
+        token = token.substring(BEARER_PREFIX_LENGTH);
```
현재 interceptor 와 리졸버가 동일한 로직을 수행하고 있는 부분이 있는 것 같아요.

- interceptor: **토큰 파싱 -> 토큰 유효성 검사** -> 인증 실패 처리 / 인증 성공
- 리졸버: **토큰 파싱 -> 토큰 유효성 검사** -> 토큰에서 파싱한 회원번호로 회원 조회

interceptor 를 거쳐온 요청이 토큰 파싱과 유효성 검사로 추가로 하는 이유가 궁금합니다💭

Profile-exe replied:
해당 피드백을 보고 `interceptor`와 `resolver`의 요청 처리 순서에 대해 알아보았습니다. `interceptor`의 `preHandle()`이 먼저 실행되며 여기서 토큰 유효성 검사를 하기 때문에 `resolver`에서는 추출한 토큰을 신뢰하고 바로 사용해도 된다는 것을 알게되었습니다. 

중복된 검증 로직을 최소화하니 코드도 깔끔해졌습니다. 감사합니다!

해당 내용은 다음 커밋들에 반영했습니다.

[1cfb8a9b](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/407/commits/1cfb8a9b6f4adf7e40031a8cd9c1ec85aa8f3230) - 토큰 검증 로직 위임
- 기존에 존재하는 함수를 재활용하여 검증으로 인한 중복 코드를 줄였습니다.

[cf0192bb](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/407/commits/cf0192bb9677a400f75c2781cd53a8c4301ff6a9) - `resolver` 검증 로직 제거 및 리펙토링
- `interceptor`를 거친 이후의 로직이므로 검증 과정을 제거했습니다.
-  `토큰 추출 -> 아이디 추출` 과정이 서로 다른 두 곳에서 중복되어 하나의 메서드(`extractMemberId`)로 축약했습니다.

hocaron replied:
코드가 더 깔끔해졌네요👍 좋습니다!

hocaron commented on src/main/java/gift/auth/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,45 @@
+        token = token.substring(BEARER_PREFIX_LENGTH);
+        Long memberId = authTokenGenerator.extractMemberId(token);
+
+        return memberRepository.findMemberByIdOrThrow(memberId);
```
현재 Member 엔티티를 그대로 반환하고 있는 것 같아요. 엔티티에 변경사항이 생겼을 때, 리졸버에서 사용하는 객체에도 영향이 전파될 것 같은데 명덕님은 어떻게 생각하시는지 궁금해요😀

Profile-exe replied:
요청과 응답에 대해 엔티티에 의존하지 않아야 하는 이유 중 하나가 **엔티티가 변경되면 요청과 응답 명세가 바뀌는** 사이트이펙트가 발생하기 때문임을 알고 있었습니다.

리졸버에서 사용하는 객체 또한 이와 같은 맥락으로 독립적인 객체를 사용해야 함을 이번 피드백을 통해 배우게 되었습니다. 감사합니다!

해당 피드백은 [b8db0fec](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/407/commits/b8db0fec72ac555ddd3a78f47efb80e3a7e2497a) 커밋에 반영했습니다.

hocaron commented on src/main/java/gift/wishlist/repository/WishListRepository.java:
```diff
@@ -0,0 +1,128 @@
+                    .param(wishListId)
+                    .update();
+        } catch (Exception e) {
+            throw WishListUpdateException.EXCEPTION;
```
현재 persist 레이어에서 에러처리를 하고 있는 부분이 많이 보입니다. 해당 에러처리를 persist 레이어에서 하는 이유가 궁금합니다💭

Profile-exe replied:
피드백을 받고 고민을 해보았는데요, 체크 예외를 받아서 비즈니스 예외로 바꾸어 던지는 로직 또한 비즈니스 로직의 일부라는 생각이 들었습니다.

컨트롤러나 리파지토리보다 서비스에서 수행하는 것이 가장 적절하다고 판단해 에러처리 로직을 서비스로 옮겼습니다.
옮기고 나니 리파지토리 코드가 순수하게 데이터를 가져오고 조작하는 코드만 남아서 보기 편했습니다.

해당 피드백은 [01163137](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/407/commits/011631373953287946819fb8bae567296b11c3bf) 커밋에 반영했습니다!

1학기 미니 과제 수행시 indent가 3을 넘지 않아야한다는 규칙이 있었습니다. 이번 과정에서도 이 규칙을 지키고싶어서 indent가 3을 넘는 부분을 `addQuantity()` 메서드로 추출했습니다.

```java
public void addWishList(Long memberId, WishListReqDto wishListReqDto) {
    // 이미 위시 리스트에 있는 상품이면 수량을 더한다.
    if (wishListRepository.isWishListExistByMemberIdAndProductId(memberId, wishListReqDto.productId())) {
        addQuantity(memberId, wishListReqDto.productId(), wishListReqDto.quantity());
        return;
    }

    try {
        wishListRepository.addWishList(memberId, wishListReqDto.productId(), wishListReqDto.quantity());
    } catch (Exception e) {
        throw WishListCreateException.EXCEPTION;
    }
}

private void addQuantity(Long memberId, Long productId, Integer quantity) {
    try {
        wishListRepository.addQuantityByMemberIdAndProductId(memberId, productId, quantity);
    } catch (Exception e) {
        throw WishListUpdateException.EXCEPTION;
    }
}
```

hocaron commented on src/main/java/gift/wishlist/service/WishListService.java:
```diff
@@ -0,0 +1,57 @@
+
+    public void addWishList(Long memberId, WishListReqDto wishListReqDto) {
+        // 이미 위시 리스트에 있는 상품이면 수량을 더한다.
+        if (wishListRepository.isWishListExistByMemberIdAndProductId(memberId, wishListReqDto.productId())) {
+            wishListRepository.updateWishListByMemberIdAndProductId(memberId, wishListReqDto.productId(), wishListReqDto.quantity());
+            return;
+        }
```
엣지 케이스에 대한 분기처리 좋습니다👍

hocaron commented on src/main/java/gift/wishlist/service/WishListService.java:
```diff
@@ -0,0 +1,57 @@
+        validateWishListExistsById(wishListId);
+
+        // 수량이 0이면 삭제
+        Integer quantity = wishListReqDto.quantity();
+        if (quantity == 0) {
+            wishListRepository.deleteWishListById(wishListId);
```
케이스별로 분기처리 로직 좋습니다👍

hocaron commented on src/main/resources/data.sql:
```diff
@@ -20,4 +20,28 @@ INSERT INTO member (name, email, password, role)
\ No newline at end of file
+       ('user3', 'user3@email.com', 'password3', 'USER');
+
+-- 위시리스트 데이터 입력: 사용자 3명이 각각 3개씩 상품을 담음
```
적절한 주석사용 좋습니다👍

hocaron commented on src/main/java/gift/wishlist/controller/WishListController.java:
```diff
@@ -0,0 +1,55 @@
+        return ResponseEntity.ok("상품을 장바구니에 담았습니다.");
+    }
+
+    @PutMapping("/{wish-list-id}")
+    public ResponseEntity<String> updateWishList(
+            @PathVariable("wish-list-id") Long wishListId,
+            @RequestBody WishListReqDto wishListReqDto
+    ) {
+        wishListService.updateWishListById(wishListId, wishListReqDto);
+        return ResponseEntity.ok("담긴 상품의 수량을 변경했습니다.");
+    }
```
[반영하지 않으셔도 좋습니다😀]
변경할 위시리스트의 회원번호와 요청한 회원번호와 일치할 때만 변경 가능하도록 하는 로직을 추가해도 좋을 것 같아요.

Profile-exe replied:
가장 중요한 부분을 놓쳤던 것 같습니다. 조회는 본인의 상품을 불러오니 사용자가 해당 상품들에 대해서만 수정, 삭제 요청을 할 것이라 생각했던 것 같습니다. 외부에서 요청만 할 수 있다면 `id`만 가지고 본인의 상품이 아니더라도 수정, 삭제할 수 있는 여지가 있다고 생각합니다.

해당 피드백은 [81da41ba](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/407/commits/81da41ba381f6ca92cf0753851da55bdbf257f41) 커밋에 반영했습니다!

예외처리의 경우 본인의 장바구니에 해당 상품이 없는 것이니 `WishListNotFoundException`을 그대로 사용했는데, 

- 존재하지 않는 위시 리스트 항목
  - NOT_FOUND: "위시 리스트를 찾을 수 없습니다."

- 존재하는 위시 리스트 항목이지만 본인의 소유가 아님
  - BAD_REQUEST: "본인의 위시 리스트만 삭제할 수 있습니다."

이런식으로 분기를 두고 예외를 나누어야 하는지 궁금합니다.

hocaron replied:
저는 분기를 따로 두지 않아도 될 것 같습니다😀(지금 상태 유지)

정상적인 유저 플로우에서는 본인의 위시리스트만 접근하는 경우가 생길텐데, 다른 회원의 위시리스트에 접근하는 경우(비정상적인/악의적인 접근)을 위한 에러케이스 및 응답을 나누지 않아도 될 것 같아요🧐

Profile-exe replied:
답변 감사드립니다! 🙇‍♂️

hocaron reviewed (changes requested):
명덕님, 안녕하세요!

3단계까지 진행하느라고 고생 많으셨습니다👍 
코드 퀄리티가 점점 좋아지는 것 같아요.
같이 작업하는 입장에서 궁금한 부분이 생겨 코멘트를 드린 부분이 있는데 확인 부탁드릴게요!

hocaron commented:
명덕님, 요청 전후처리시에 인터셉터, 필터, AOP 중에 어디가 적절한지 고민을 많이 하게 되는 것 같아요!

심심할 때 보기 스택오버플로우 링크 공유드립니다!
https://stackoverflow.com/questions/35856454/difference-between-interceptor-and-filter-in-spring-mvc

hocaron commented:
1. 위시리스트 추가 로직에서 기존에 상품이 존재하는지를 분기를 두어 확인했습니다. 추가 요청 대상인 상품이 이미 존재한다면 기존의 수량에서 요청된 수량만큼 더 증가시키는 방식으로 구현했습니다.

이러한 로직을 리파지토리에서 하나의 메서드로 처리할 수 있을텐데 서비스에 구현하는게 올바른 방식인지 궁금합니다. 제 생각에는 서비스에서 비즈니스 로직을 책임지고 리파지토리는 데이터를 조회, 조작하는 책임만 있다고 생각해 서비스에서 분기를 두고 리파지토리에는 각 분기에 해당하는 메서드를 따로 만들었습니다.

하나의 메서드라면 어떻게 구현할 수  있을까요?!😀

Profile-exe committed:
fix(JwtTokenProvider): validateToken 메서드 검증 방식 변경

반환된 `Claims`가 `null`이거나 예외가 발생하면 `false`를 반환하도록 변경

Profile-exe committed:
refactor(JwtInterceptor): 토큰 검증 로직 위임

Profile-exe committed:
refactor(AuthTokenGenerator): extractToken 제거 및 extractMemberId에 토큰 추출 로직 추가

`extractMemberId`를 사용 하는 코드에서 항상 `extractToken`을 사용하는 것이 발견되어 `jwtTokenProvider`의 `getToken`과 `getMemberId`를 한 메서드에서 호출하도록 변경

이를 통해 `extractMemberId`라는 메서드 하나만 호출하게 되어 코드가 간결해짐

Profile-exe committed:
refactor(ArgumentResolver): Member 엔티티 의존성을 제거하기 위해 DTO로 반환

Profile-exe committed:
style(WishListRepository): 불필요한 빈 줄 제거

Profile-exe committed:
feat: 본인의 장바구니에 있는 상품만 수정, 삭제 가능하도록 변경

Profile-exe committed:
refactor: 비즈니스 예외 던지는 로직을 서비스에서 수행하도록 변경

Profile-exe committed:
feat(WishListService): update에도 검증 로직 추가

수량이 0인 경우 `deleteWishListById`로 넘어가며 검증을 수행하는데 수량이 0이 아닌 경우 검증 과정이 없어서 추가함

Profile-exe commented:
질문과 이에 대해 답변해주신 부분

> 이러한 로직을 리파지토리에서 하나의 메서드로 처리할 수 있을텐데 서비스에 구현하는게 올바른 방식인지 궁금합니다. 제 생각에는 서비스에서 비즈니스 로직을 책임지고 리파지토리는 데이터를 조회, 조작하는 책임만 있다고 생각해 서비스에서 분기를 두고 리파지토리에는 각 분기에 해당하는 메서드를 따로 만들었습니다.
> 
> 하나의 메서드라면 어떻게 구현할 수 있을까요?!😀

하나의 메서드에 둔다면 `WishListRepository`에서 
```java
public void addWishList(Long memberId, Long productId, Integer quantity) {
    if (isWishListExistByMemberIdAndProductId(memberId, productId)) {
        var sql = """
            update wish_list
            set quantity = quantity + ?
            where member_id = ? and product_id = ?
            """;

        jdbcClient.sql(sql)
                .param(quantity)
                .param(memberId)
                .param(productId)
                .update();
        return;
    }

    var sql = """
                insert into wish_list (member_id, product_id, quantity)
                values (?, ?, ?)
                """;

    jdbcClient.sql(sql)
            .param(memberId)
            .param(productId)
            .param(quantity)
            .update();
}
```

이런식으로 구현해 서비스에서는 메서드 하나를 호출할 수 있지 않을까 생각했습니다!

hocaron commented:
> 하나의 메서드에 둔다면 `WishListRepository`에서
> 
> ```java
> public void addWishList(Long memberId, Long productId, Integer quantity) {
>     if (isWishListExistByMemberIdAndProductId(memberId, productId)) {
>         var sql = """
>             update wish_list
>             set quantity = quantity + ?
>             where member_id = ? and product_id = ?
>             """;
> 
>         jdbcClient.sql(sql)
>                 .param(quantity)
>                 .param(memberId)
>                 .param(productId)
>                 .update();
>         return;
>     }
> 
>     var sql = """
>                 insert into wish_list (member_id, product_id, quantity)
>                 values (?, ?, ?)
>                 """;
> 
>     jdbcClient.sql(sql)
>             .param(memberId)
>             .param(productId)
>             .param(quantity)
>             .update();
> }
> ```
> 
> 이런식으로 구현해 서비스에서는 메서드 하나를 호출할 수 있지 않을까 생각했습니다!

아핫! 아주 좋은 고민이라고 생각합니다. 명덕님께서 말씀주신 것처럼 서비스에서 분기를 두고 리파지토리에는 각 분기에 해당하는 메서드를 따로 만드는게 역할 분리 측면에서 좋을 것 같습니다.

코드 재사용성 측면도 생각해볼 수 있을 것 같아요.

서비스 메서드에 대응하는 레포지토리 메서드를 하나씩 만들게 되면 해당 서비스 로직에서만 사용할 수 밖에 없고, 아주 조금의 서비스 로직이 다른 경우에도 레포지토리 메서드를 재사용하는 것 대신 새로 만들어야 하는 경우가 발생할 수 있는데 이때 중복되는 코드가 생기고, 관리 포인트도 늘 것 같아요🤔

hocaron reviewed (approved):
명덕님, 2주차 과제 수행하시느라 고생 많으셨습니다💪
코드 퀄리티가 점점 높아지는 것 같습니다! 코드 작성하시면서 고민하셨던 부분 공유해주셔서 감사합니다.

코멘트에 대한 부분이 모두 해결되어서 PR 은 머지하려고 합니다!
고민이나 궁금한 부분은 코멘트나 DM 주시면 확인하겠습니다🙇‍♀️

Profile-exe commented:
> 아핫! 아주 좋은 고민이라고 생각합니다. 명덕님께서 말씀주신 것처럼 서비스에서 분기를 두고 리파지토리에는 각 분기에 해당하는 메서드를 따로 만드는게 역할 분리 측면에서 좋을 것 같습니다.
> 
> 코드 재사용성 측면도 생각해볼 수 있을 것 같아요.
> 
> 서비스 메서드에 대응하는 레포지토리 메서드를 하나씩 만들게 되면 해당 서비스 로직에서만 사용할 수 밖에 없고, 아주 조금의 서비스 로직이 다른 경우에도 레포지토리 메서드를 재사용하는 것 대신 새로 만들어야 하는 경우가 발생할 수 있는데 이때 중복되는 코드가 생기고, 관리 포인트도 늘 것 같아요🤔

역할 분리 측면 뿐만 아니라 코드 재사용성 측면에서도 유용하군요. 좋은 인사이트 감사드립니다! 메서드를 만들 때 이러한 부분들을 고려해보겠습니다
